packages/serverless/src/apollo/app/merkle-tree.ts

```diff
@@ -0,0 +1,173 @@
+import { Field } from 'o1js';
+import Joi from 'joi';
+import GraphQLJSON from 'graphql-type-json';
+import { BigIntResolver } from 'graphql-scalars';
+import resolverWrapper from '../validation';
+import { databaseName, indexNumber } from './common';
+import { TDatabaseRequest } from './database';
+import DistributedLock from '../../helper/distributed-lock';
+import MerkleTreeService from '../../service/MerkleTreeService';
+
+export type TMerkleTreeInitializeRequest = TDatabaseRequest & {
+  height: number;
+};
+
+export type TMerkleTreeBuildRequest = TDatabaseRequest & {
+  amount: number;
+};
+
+export type TMerkleTreeStateRequest = TDatabaseRequest & {
+  root: string;
+};
+
+export type TMerkleTreeIndexRequest = TMerkleTreeStateRequest & {
+  index: bigint;
+};
+
+export type TMerkleTreeSetLeafRequest = TMerkleTreeIndexRequest & {
+  hash: string;
+};
+
+export type TMerkleTreeGetNodeRequest = TMerkleTreeIndexRequest & {
+  level: number;
+};
+
+export const MerkleTreeInitializeRequest =
+  Joi.object<TMerkleTreeInitializeRequest>({
+    databaseName,
+    height: Joi.number().required(),
+  });
+
+export const MerkleTreeIndexRequest = Joi.object<TMerkleTreeIndexRequest>({
+  databaseName,
+  root: Joi.string().required(),
+  index: indexNumber,
+});
+
+export const MerkleTreeSetLeafRequest = Joi.object<TMerkleTreeSetLeafRequest>({
+  databaseName,
+  index: indexNumber,
+  hash: Joi.string().required(),
+});
+
+export const MerkleTreeBuildRequest = Joi.object<TMerkleTreeBuildRequest>({
+  databaseName,
+  amount: Joi.number().required(),
+});
+
+export const MerkleTreeGetNodeRequest = Joi.object<TMerkleTreeGetNodeRequest>({
+  databaseName,
+  root: Joi.string().required(),
+  index: indexNumber,
+  level: Joi.number().required(),
+});
+
+export const typeDefsMerkleTree = `#graphql
+scalar BigInt
+scalar JSON
+type Query
+type Mutation
+
+type MerkleProof {
+  isLeft: Boolean!
+  sibling: String!
+}
+
+extend type Query {
+  getNode(databaseName: String!, level: Int!, index: BigInt!): String!
+  getWitness(databaseName: String!, root: String!, index: BigInt!): [MerkleProof]!
+}
+
+extend type Mutation {
+  createMerkleTree(databaseName: String!, height: Int!): Boolean
```

Comment:
I think we shouldn't allow users to manipulate the Merkle tree themselves. Instead we all users to perform data manipulation then we handle Merkle tree manipulation for them.

Reply:
There are pretty many use cases when users need to work only with MT, aren't there?